buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -0,0 +1,1402 @@
+ * based on the id of the thread. This spreads the contention of multi-threaded access across the magazines.
+ * If contention is detected above a certain threshold, the number of magazines are increased in response to the
+ * contention.
+ * <p>
+ * The magazines maintain histograms of the sizes of the allocations they do. The histograms are used to compute the
+ * preferred chunk size. The preferred chunk size is one that is big enough to service 10 allocations of the
+ * 99-percentile size. This way, the chunk size is adapted to the allocation patterns.
+ * <p>
+ * Computing the preferred chunk size is a somewhat expensive operation. Therefore, the frequency with which this is
+ * done, is also adapted to the allocation pattern. If a newly computed preferred chunk is the same as the previous
+ * preferred chunk size, then the frequency is reduced. Otherwise, the frequency is increased.
+ * <p>
+ * This allows the allocator to quickly respond to changes in the application workload,
+ * without suffering undue overhead from maintaining its statistics.
+ * <p>
+ * Since magazines are "relatively thread-local", the allocator has a central queue that allow excess chunks from any
+ * magazine, to be shared with other magazines.
+ * The {@link #createSharedChunkQueue()} method can be overridden to customize this queue.
+ */
+@SuppressJava6Requirement(reason = "Guarded by version check")
+@UnstableApi
+final class AdaptivePoolingAllocator {
+    private static final int RETIRE_CAPACITY = 4 * 1024;
+    private static final int MIN_CHUNK_SIZE = 128 * 1024;
+    private static final int MAX_STRIPES = NettyRuntime.availableProcessors() * 2;
+    private static final int BUFS_PER_CHUNK = 10; // For large buffers, aim to have about this many buffers per chunk.
+
+    /**
+     * The maximum size of a pooled chunk, in bytes. Allocations bigger than this will never be pooled.
+     * <p>
+     * This number is 10 MiB, and is derived from the limitations of internal histograms.
+     */
+    private static final int MAX_CHUNK_SIZE =
+            BUFS_PER_CHUNK * (1 << AllocationStatistics.HISTO_MAX_BUCKET_SHIFT); // 10 MiB.
+
+    /**
+     * The capacity if the central queue that allow chunks to be shared across magazines.
+     * The default size is {@link NettyRuntime#availableProcessors()},
+     * and the maximum number of magazines is twice this.
+     * <p>
+     * This means the maximum amount of memory that we can have allocated-but-not-in-use is
+     * 5 * {@link NettyRuntime#availableProcessors()} * {@link #MAX_CHUNK_SIZE} bytes.
+     */
+    private static final int CENTRAL_QUEUE_CAPACITY = SystemPropertyUtil.getInt(
+            "io.netty.allocator.centralQueueCapacity", NettyRuntime.availableProcessors());
+
+    private final ChunkAllocator chunkAllocator;
+    private final Queue<ChunkByteBuf> centralQueue;
+    private final StampedLock magazineExpandLock;
```

Comment:
Unless I am missing something obvious, this is going to cause the class to not be loadable in Java 6 and 7 and the version check in the constructor won't help because it's executed too late.

Reply:
I think you're right. I'll put up a PR.

Reply:
👌

Reply:
#14101